lightning/src/routing/scoring.rs

```diff
@@ -460,25 +466,420 @@ impl<T: Time> Readable for ChannelFailure<T> {
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat`.
+	fn failed_at_channel(&mut self, amount_msat: u64) {
+		if amount_msat < self.max_liquidity_msat() {
+			self.set_max_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat` downstream.
+	fn failed_downstream(&mut self, amount_msat: u64) {
+		if amount_msat > self.min_liquidity_msat() {
+			self.set_min_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when successfully routing `amount_msat`.
+	fn successful(&mut self, amount_msat: u64) {
+		let max_liquidity_msat = self.max_liquidity_msat().checked_sub(amount_msat).unwrap_or(0);
+		self.set_max_liquidity_msat(max_liquidity_msat);
+	}
+
+	/// Adjusts the lower bound of the channel liquidity balance in this direction.
+	fn set_min_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.min_liquidity_offset_msat = amount_msat;
+		*self.max_liquidity_offset_msat = if amount_msat > self.max_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.max_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
+	}
+
+	/// Adjusts the upper bound of the channel liquidity balance in this direction.
+	fn set_max_liquidity_msat(&mut self, amount_msat: u64) {
+		*self.max_liquidity_offset_msat = self.capacity_msat.checked_sub(amount_msat).unwrap_or(0);
+		*self.min_liquidity_offset_msat = if amount_msat < self.min_liquidity_msat() {
+			0
+		} else {
+			self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+		};
+		*self.last_updated = self.now;
+	}
+}
+
+impl<G: Deref<Target = NetworkGraph>, T: Time> Score for ProbabilisticScorerUsingTime<G, T> {
+	fn channel_penalty_msat(
+		&self, short_channel_id: u64, amount_msat: u64, capacity_msat: u64, source: &NodeId,
+		target: &NodeId
+	) -> u64 {
+		let liquidity_penalty_multiplier_msat = self.params.liquidity_penalty_multiplier_msat;
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let success_probability = self.channel_liquidities
+			.get(&short_channel_id)
+			.unwrap_or(&ChannelLiquidity::new())
+			.as_directed(source, target, capacity_msat, liquidity_offset_half_life)
+			.success_probability(amount_msat);
+		// NOTE: If success_probability is ever change to return 0.0, log10 is undefined so return
+		// u64::max_value instead.
+		debug_assert!(success_probability > core::f64::EPSILON);
+		(-(success_probability.log10()) * liquidity_penalty_multiplier_msat as f64) as u64
+	}
+
+	fn payment_path_failed(&mut self, path: &[&RouteHop], short_channel_id: u64) {
+		let amount_msat = path.split_last().map(|(hop, _)| hop.fee_msat).unwrap_or(0);
+		let liquidity_offset_half_life = self.params.liquidity_offset_half_life;
+		let network_graph = self.network_graph.read_only();
+		let hop_sources = path.iter().map(|hop| NodeId::from_pubkey(&hop.pubkey));
```

Comment:
Instead of getting the source PK from the path, let's get it as a side-effect of `channel.as_directed_to`, that way we'll score public first-hop channels as well.